### DISCUSSION TITLE
Data corrupted when using wire:model

### DISCUSSION THREAD

**Discussion by rafa-acioly** [livewire/livewire#3053] [7. Bugs (archived)]
### Description
I'm trying do create a dynamic 3-level dependent dropdowns and when i try to use `wire:model` on one of this dropdowns i got the error saying:

> Livewire encountered corrupt data when trying to hydrate the [product-categories] component.

## The template

```php
<!-- more HTML -->

    <x-jet-dialog-modal>
        <x-slot name="title">Category settings</x-slot>
        <x-slot name="content">
            <label class="text-gray-700">
                Category
                <select wire:model="firstLevel">
                    <option disabled selected value="">Select...</option>
                    @foreach($meliCategories as $category)
                        <option value="{{ $category->getId() }}">{{ $category->getName() }}</option>
                    @endforeach
                </select>
            </label>

            @if(!is_null($secondLevel))
                <label class="text-gray-700 hidden">
                    Category
                    <select>
                        <option disabled selected value="">Select...</option>
                    </select>
                </label>
            @endif

            @if(!is_null($thirdLevel))
                <label class="text-gray-700 hidden">
                    Category
                    <select>
                        <option disabled selected value="">Select...</option>
                    </select>
                </label>
            @endif
        </x-slot>
        <x-slot name="footer">
            <x-jet-secondary-button wire:loading.attr="disabled">
                Nevermind
            </x-jet-secondary-button>
        </x-slot>
    </x-jet-dialog-modal>

<!--- more HTML rendering the "categories" attribute in a table -->
```

## The Component
```php
class ProductCategories extends Component
{
    public $categories = [];

    public $meliCategories = [];
    public $secondLevelCategory = [];
    public $thirdLevelCategory = [];

    public $firstLevel = null;
    public $secondLevel = null;
    public $thirdLevel = null;

    public $editModal = null;

    /**
     * @var CategoryService|mixed
     */
    protected $service;

    public $confirmFullSync;

    public function mount()
    {
        $this->service = new CategoryService(new MeliAdapter(env('MELI_ID'), env('MELI_SECRET')));

        $this->meliCategories = $this->service->findCategories(Site::BRASIL)->toArray(); // array of custom Category class

        $this->categories = Auth::user()->productCategories;
        $this->secondLevelCategory = collect();
        $this->thirdLevelCategory = collect();
    }

    public function render()
    {
        return view('settings.categories.list');
    }

    public function syncAll()
    {
        $credential = Auth::user()->credential;
        ProductCategoriesSync::dispatch(new Woocommerce($credential), Auth::user());

        $this->confirmFullSync = !$this->confirmFullSync;
        notify()->success('Success', 'It works');
    }

    public function showEditModal($category)
    {
        $this->editModal = $category;
    }

    public function updatedFirstLevelCategory($data)
    {
        dd($data);
    }
}
```

**Context**
- Livewire version: `^2.0`
- Laravel version: `^8.0`
- Alpine version: `2.6.0`
- Browser: Chrome

**Comment by rafa-acioly** [comment]
I've started from the begining and i notica that if i use a third-party lib to load the data from a API the component brake.

I tried to add this into the `mount` method

```php
class ProductCategories extends Component
{
    protected $service;

    public $confirmFullSync, $editModal;

    public $categories;
    
    private $categoriesOption; // <-- if i use this as public the error shows up

    public function mount(CategoryService $service)
    {
        $this->categories = Auth::user()->productCategories;
        $this->categoriesOption = $service->findCategories(Site::BRASIL)->toArray(); // Array of Category class
    }
}
```

The class from the package is:

> https://github.com/discovery-tecnologia/dsc-mercado-livre/blob/2320cdacee6a67b5e64a275280193c0d62df3406/src/Requests/Category/CategoryService.php#L13


After adding this line if i try to sendo a "wire:click" event i got the error saying that `Livewire encountered corrupt data`

**Comment by pkeogan** [comment]
I was having a similar issue last week.

On our dev servers, the Component would function as designed, but when deployed to prod it would throw a `Livewire encountered corrupt data when trying to hydrate the [***] component.` error. 

I never was able to pinpoint the exact location of the problem, but was able to fix it. My assumption was during the livewire lifecycle, of `dehydrate` and `hydrate` Laravel's blade cache features, was somehow modifying code.

This is what my code looked like prior:
```
---DraftPoint.php
...
public $point;
public $recruits;
...
public function Mount(Point $point){
  $this->point = $point;
  $this->recruits= User::all();
}
...
```
```
---draft-point.blade.php
...
  <x-input.select
          wire:model="point.recruit_id"
          required=true
          :options="$recruits->pluck('name', 'id')->toArray()"
          />
...
```

Then after changing the following code
```
---DraftPoint.php
...
public $point;
public $recruits;
...
public function Mount(Point $point){
  $this->point = $point;
  $this->recruits= User::all()->pluck('name', 'id')->toArray();
}
...
```
```
---draft-point.blade.php
...
  <x-input.select
          wire:model="point.recruit_id"
          required=true
          :options="$recruits"
          />
...
```
The issue had something to do with modification of the collection on the blade file, and Laravel's caching methods for blade files. 

Sorry if that doesn't help!

**Comment by codewithfeeling** [comment]
This is my life right now.

I'm loading some objects via an API call, then making an array of tags from the retrieved data. This is used to populate the options of a `<select>`

````
<select wire:model="filterType">
    <option value="">All types</option>
    @foreach($types as $type)
        <option value="{{ $type }}">{{ $type }}</option>
    @endforeach
</select>
````

Every time I select a value, I get the `Livewire encountered corrupt data when trying to hydrate...` error. Adding `.lazy` to the wire:model has no effect.

I can't see what I'm doing wrong but it is making me want to poke my eyeballs out with a herring.

**Reply by codewithfeeling** [reply]
I discovered what I was doing wrong just in case it helps anyone else.

When creating the tags, I was using Laravel's Collection methods, but forgot to convert `toArray()` at the end. This meant that the `$types` variable initially set to an empty array became a Collection which is not valid.

Simply adding `->toArray()` at the end of the collection methods sorted this and it works now.

However, no sooner have I solved that than I add a simple checkbox to the rendered tag cloud:

````
<label class="inline-block mb-4">
    <input type="checkbox" wire:model="hideSingleUseTags" /> Hide single-use tags
</label>
````

And I'm back to square one... `Livewire encountered corrupt data when trying to hydrate...`

it's pretty exhausting.

**Comment by ChrisSantiago82** [comment]
Once I had a similar problem. Everything worked until we pushed it to production.
Then we noticed that one model had a value of -0. This is valid in php but gets changed to 0 in JavaScript. And when it's being send back from JavaScript to php, the data was "corrupted" because JavaScript changed it from -0 to 0. Maybe you are having a similar problem?

**Comment by debjit** [comment]
Ok, so I have one question, do you/anyhow changing the database? I have the problem when I was deleting the database part livewire using.

**Reply by ChrisSantiago82** [reply]
The issue is, you could have a value in your database/model that causes this issue.
You could have a value, that is fine with php but then it's being changed by JavaScript and livewire then thinks somebody has tampered with that data.

**Reply by ChrisSantiago82** [reply]
I had once a case with -0 and 0 and it took me a long time to find the issue.
But it's just and idea
